src/atomic.cr

```diff
@@ -264,8 +264,8 @@ struct Atomic(T)
   # ```
   def swap(value : T, ordering : Ordering = :sequentially_consistent)
     {% if T.union_types.all? { |t| t == Nil || t < Reference } && T != Nil %}
-      address = atomicrmw(:xchg, pointerof(@value).as(LibC::SizeT*), LibC::SizeT.new(value.as(Void*).address), ordering)
-      Pointer(T).new(address).as(T)
+      address = atomicrmw(:xchg, pointerof(@value).as(typeof(@value.object_id)*), value.object_id, ordering)
+      Pointer(Void).new(address).as(T)
```

Comment:
Unless I'm mistaken `#object_id` is _always_ casted to 64-bit, whatever of the architecture (32 or 64-bit) which would lead to an expensive DWCAS on i686 and it's disabled by default by LLVM. I'm not sure about ARM.

I'd keep the cast as SizeT.

Reply:
Hm, good point. Maybe we would want `#object_id` to actually be of platform-specific width?
Crystal itself just doesn't have a `SizeT` type yet, so it's just been the default `UInt64` for now.

However, thinking a bit more about this `#object_id` could be overriden by any type (for whatever reason). So taking the address directly is actually a bit more robust.

Reply:
Yeah, the return type of `Pointer#address` is documented as `UInt64`. But I believe the underlying `LLVMBuildPtrToInt` actually casts to an integer of the same width as a pointer.

Reply:
`LLVMBuildPtrToInt` takes a target type, which is currently hardcoded to `i64`:

https://github.com/crystal-lang/crystal/blob/723f8296370c3c77efb6770eb7480d88c8b4d98f/src/compiler/crystal/codegen/primitives.cr#L796-L798